### ISSUE TITLE
React Native Style Editor UX issues

### ISSUE THREAD

**Issue by soul-wish** [facebook/react-devtools#668]
Hi, everyone! 
As a follow-up to conversation in #627 

---

1. There are some inconsistencies in style editor UI (some paddings, input's layout etc.). But I've checked other issues/PR's in the repo and saw that all these issues will be fixed in the nearest future;

---

2. It would be really cool to be able to navigate to the style value by pressing tab key on the keyboard. Now there is only one input field for the new style key, and only after blur event second field appears. So right now there is no way to add new styles by keyboard only - you should always use your touchpad or mouse;

![blur](https://cloud.githubusercontent.com/assets/1968098/25543920/0df7332e-2c61-11e7-8aa2-35e22c87a0a4.gif)

---

3. If I got it correctly, `RCTView` is always a child of `View` so it inherits all its styles (correct me if I'm wrong). And I've noticed that you can not simply edit `RCTView` styles because they will be reset automatically from the parent `View`. So, maybe it would be better just to display all these styles and not allow to change them (readonly fields, or even better - some type of list with style's keys and values);

---

4. As for the `View`s styles I've found few strange issues: 
    - I can not delete original styles (they just keep reappearing, but maybe it should work in this way);
![original-styles](https://cloud.githubusercontent.com/assets/1968098/25543945/22defc86-2c61-11e7-960b-590812b134a3.gif)

    - But I can delete the styles that were added by me manually (with a small fix in this PR);
![new-styles](https://cloud.githubusercontent.com/assets/1968098/25543979/3dc8335a-2c61-11e7-9121-a2900782c665.gif)

    ---

    - I can not change the keys of the styles (but I can change their values);
![original-styles-values-keys](https://cloud.githubusercontent.com/assets/1968098/25543987/45660920-2c61-11e7-8127-5efc1a7e0243.gif)

    ---

    - If I try to change the key of the original style it will be added as a new style record (key->value) at the bottom of styles editing form;
![original-styles-edit-key](https://cloud.githubusercontent.com/assets/1968098/25543993/4fd625b6-2c61-11e7-87a4-d7da121bdbc1.gif)

    ---

5. I'm not sure if it is possible to not add new style if the key is incorrect? Right now it works in this way:

![incorrect-keys](https://cloud.githubusercontent.com/assets/1968098/25543998/558a7804-2c61-11e7-96f9-24628fc13445.gif)

---

PS. I apologize in case I miss some important points or concepts (as I said earlier - I have zero experience in RN) ;)

**Comment by gaearon** [comment]
@ansumanshah You did great work on https://github.com/facebook/react-devtools/pull/671, would you be interested in taking a look at these issues?

**Comment by ansumanshah** [comment]
@gaearon I would be happy to take it

**Comment by gaearon** [comment]
Awesome, thanks!

**Comment by gaearon** [comment]
>I can not delete original styles (they just keep reappearing, but maybe it should work in this way);

I think the real problem here is that our [style setting logic](https://github.com/facebook/react-devtools/blob/dc8f87d51c47aa27c79207a65b8c2f2adf5f6730/plugins/ReactNativeStyle/setupBackend.js#L141) probably works by shallow merging, so if your style was `{a: 1, b: 2}` and you edited it to be `{a: 1}`, the end result is still `{a: 1, b: 2}`. The fix would be to set `undefined` instead of performing real deletions, so that it ends up being `{a: 1, b: undefined}`, and filter `undefined` values from the UI.

**Comment by ansumanshah** [comment]
> I can not change the keys of the styles (but I can change their values)

If we leave key or value as an empty string, the value is set as `undefined` and it is removed.

> I can not delete original styles (they just keep reappearing, but maybe it should work in this way)

The main issue with deleting the original styles is something else.

[`data.props.style`](https://github.com/facebook/react-devtools/blob/master/plugins/ReactNativeStyle/setupBackend.js#L106) actually returns the ID and not the style object which we can only get after we resolve/flatten it with [resolveRNStyle](https://github.com/facebook/react-devtools/blob/master/plugins/ReactNativeStyle/setupBackend.js#L59)
so the rest of the shallowClone fn and deleting the old key doesn't actually work.

Moreover in case of [setStyle](https://github.com/facebook/react-devtools/blob/master/plugins/ReactNativeStyle/setupBackend.js#L158) this moves to the else part and adds an array instead.

@gaearon was the rename function working previously or something changed in the API? what is the ideal solution to this?

**Comment by gaearon** [comment]
I don't know if it ever worked. Can we treat rename as deleting a key and adding a new key? I'm not sure I understand what makes it harder.

>and how should we deal with Incorrect Key?

I think it's not too big of a deal given that React itself will warn you with a yellow box.

**Comment by ansumanshah** [comment]
Yes, we can delete the key in the frontend and a new one.

It's just that I am not sure if that is the best solution.

**Comment by gaearon** [comment]
Sounds okay to me.

**Comment by bvaughn** [comment]
React DevTools has been rewritten and [recently launched a new version 4 UI](https://reactjs.org/blog/2019/08/15/new-react-devtools.html). The source code for this rewrite was done in a separate repository and now lives in the main React repo ([github.com/facebook/react](https://github.com/facebook/react/)).

Because version 4 was a total rewrite, and all issues in this repository are related to the old version 3 of the extension, I am closing all issues in this repository. If you can still reproduce this issue, or believe this feature request is still relevant, please open a new issue in the React repo: https://github.com/facebook/react/issues/new?labels=Component:%20Developer%20Tools
